### ISSUE TITLE
SV-DKL Approach / uncertainty measurements

### ISSUE THREAD

**Issue by cherepanovic** [cornellius-gp/gpytorch#1074]
what is the way to measure the uncertainty from the output beside the output probability? 

Is the variance from the SoftmaxLikelihood a proper way? ([SV-DKL example](https://github.com/cornellius-gp/gpytorch/blob/master/examples/06_PyTorch_NN_Integration_DKL/Deep_Kernel_Learning_DenseNet_CIFAR_Tutorial.ipynb))

thank a lot :pray:

**Comment by gpleiss** [comment]
This depends on the type of uncertainty you want to measure. If you want to measure "model uncertainty" (e.g. the "epistemic uncertainty") then look at the variance of the latent functions (i.e. the output from the GP before the likelihood). If you want to measure "data uncertainty" (e.g. the "aleatoric uncertainty") then look at e.g. the entropy of the softmax distributions after the output of the likelihood.

This paper does a good job differentiating between the two types of uncertainty: https://arxiv.org/abs/1703.04977

**Comment by cherepanovic** [comment]
`gpytorch.likelihoods.SoftmaxLikelihood(num_features=model.num_dim, num_classes=num_classes)`

provides already the SoftmaxLikelihood. In this case, there is no need to do something additionally. 

```
m_output = model(input_ids)
output = likelihood(m_output)

print(output.probs.mean(0))
```

output.probs.mean(0) returns already the aleatoric uncertainty.

Is my assumption correct?

**Comment by cherepanovic** [comment]
>If you want to measure "model uncertainty" (e.g. the "epistemic uncertainty") then look at the variance of the latent functions (i.e. **the output from the GP before the likelihood**). 

**the output from the GP before the likelihood** is the `m_output = model(input_ids)` which is a `MultitaskMultivariateNormal` with not the same dimension as the number of labels. 

How to infer the epistemic uncertainty from MultitaskMultivariateNormal is not clear to me (?)

Thank a lot!

**Comment by gpleiss** [comment]
It depends how you want to quantify "uncertainty." From the GP's MultitaskMultivariateNormal (for epistemic uncertainty) - you could measure the average variance across tasks or the entropy of the distribution. From the likelihood output (which combines epistemic and aleatoric uncertainty) you could look at the maximum confidence (i.e. the highest logit) or again the entropy

**Comment by cherepanovic** [comment]
> output.probs.mean(0) returns already the aleatoric uncertainty.

this would be the highest logit, which combines epistemic and aleatoric uncertainty, is it right?

I need the uncertainties in prediction for an instance, therefore, I didn't get the points regarding such as 
> you could measure the average variance across tasks or the entropy of the distribution

and

>  again the entropy (from the likelihood output)

Can you give me some hints 🙏

**Comment by gpleiss** [comment]
`output.probs.mean(0)` gives you all the logits (for all classes). For computing these various distribution metrics use [some of PyTorch's built in distribution methods]( https://pytorch.org/docs/stable/distributions.html#torch.distributions.multivariate_normal.MultivariateNormal)

**Comment by cherepanovic** [comment]
>the likelihood output (which combines epistemic and aleatoric uncertainty) look at the maximum confidence (i.e. the highest logit)

**which combines epistemic and aleatoric uncertainty** what are the points which prove that it is a combination of epistemic and aleatoric uncertainties? Thanks!
